python/lsst/daf/butler/registry/interfaces/_database.py

```diff
@@ -468,29 +455,126 @@ def temporary_table(
+        connection and whether this is a new outermost transaction.
+
+        This method is for internal `Database` calls that need the actual
+        SQLAlchemy connection object.  It should be overridden by subclasses
+        instead of `transaction` itself.
+
+        Parameters
+        ----------
+        interrupting : `bool`, optional
+            If `True` (`False` is default), this transaction block may not be
+            nested without an outer one, and attempting to do so is a logic
+            (i.e. assertion) error.
+        savepoint : `bool`, optional
+            If `True` (`False` is default), create a `SAVEPOINT`, allowing
+            exceptions raised by the database (e.g. due to constraint
+            violations) during this transaction's context to be caught outside
+            it without also rolling back all operations in an outer transaction
+            block.  If `False`, transactions may still be nested, but a
+            rollback may be generated at any level and affects all levels, and
+            commits are deferred until the outermost block completes.  If any
+            outer transaction block was created with ``savepoint=True``, all
+            inner blocks will be as well (regardless of the actual value
+            passed).  This has no effect if this is the outermost transaction.
+        lock : `Iterable` [ `sqlalchemy.schema.Table` ], optional
+            A list of tables to lock for the duration of this transaction.
+            These locks are guaranteed to prevent concurrent writes and allow
+            this transaction (only) to acquire the same locks (others should
+            block), but only prevent concurrent reads if the database engine
+            requires that in order to block concurrent writes.
+        for_temp_tables : `bool`, optional
+            If `True`, this transaction may involve creating temporary tables.
+
+        Returns
+        -------
+        context : `AbstractContextManager` [ `tuple` [ `bool`,
+                `sqlalchemy.engine.Connection` ] ]
+            A context manager that commits the transaction when it is exited
+            without error and rolls back the transactoin when it is exited via
+            an exception.  When entered, it returns a tuple of:
+
+            - `is_new` (`bool`) whether this is a new (outermost) transaction;
+            - `connection` (`sqlalchemy.engine.Connection`) the connection.
+        """
+        with self._session() as connection:
+            already_in_transaction = connection.in_transaction()
+            assert not (interrupting and already_in_transaction), (
+                "Logic error in transaction nesting: an operation that would "
+                "interrupt the active transaction context has been requested."
+            )
+            savepoint = savepoint or connection.in_nested_transaction()
+            trans: sqlalchemy.engine.Transaction
```

Comment:
Does this need `| None`?

Reply:
Yup, good catch.  I'm sure MyPy is not complaining only because it sees all the SQLAlchemy types as `Any` (for now).